File: dynamic_programming/fibonacci.cpp
#include <iostream>
#include <vector>

class Fibonacci {
private:
  std::vector<int> sequence;

public:
  Fibonacci() {
    sequence.push_back(0);
    sequence.push_back(1);
  }

  std::vector<int> get(int index) {
    if (index <= 0)
      return {};

    int difference = index - (sequence.size());
    if (difference >= 0) {
      for (int i = 0; i <= difference; ++i) { // Calculate enough
        sequence.push_back(sequence.back() + sequence[sequence.size() - 2]);
      }
    }

    // Return sub-vector
    std::vector<int> result;
    for (int i = 0; i < index && i < sequence.size(); ++i) {
      result.push_back(sequence[i]);
    }
    return result;
  }
};

int main() {
  Fibonacci fib;
  std::vector<int> res = fib.get(10);
  for (int num : res)
    std::cout << num << " ";
  std::cout << std::endl;
  return 0;
}

File: dynamic_programming/fibonacci.js
/**
 * Fibonacci sequence using Dynamic Programming
 */
class Fibonacci {
    constructor() {
        this.sequence = [0, 1];
    }

    get(index) {
        if (index <= 0) return [];

        const currentLength = this.sequence.length;
        if (index > currentLength) {
            for (let i = currentLength; i < index; i++) {
                this.sequence.push(this.sequence[this.sequence.length - 1] + this.sequence[this.sequence.length - 2]);
            }
        }
        return this.sequence.slice(0, index);
    }
}

const fib = new Fibonacci();
console.log(fib.get(10));
